numpy/core/src/multiarray/methods.c

```diff
@@ -63,34 +63,48 @@ NpyArg_ParseKeywords(PyObject *keys, const char *format, char **kwlist, ...)
 
 
 /*
- * Forwards an ndarray method to a the Python function
- * numpy.core._methods.<name>(...)
+ * Forwards a method call to a Python function while adding `self`:
+ * callable(self, ...)
  */
 static PyObject *
-forward_ndarray_method(PyArrayObject *self, PyObject *args, PyObject *kwds,
-                            PyObject *forwarding_callable)
+npy_forward_method(
+        PyObject *callable, PyObject *self,
+        PyObject *const *args, Py_ssize_t len_args, PyObject *kwnames)
 {
-    PyObject *sargs, *ret;
-    int i, n;
+    PyObject **new_args;
 
-    /* Combine 'self' and 'args' together into one tuple */
-    n = PyTuple_GET_SIZE(args);
-    sargs = PyTuple_New(n + 1);
-    if (sargs == NULL) {
-        return NULL;
+    /*
+     * Unfortunately, `PY_VECTORCALL_ARGUMENTS_OFFSET` seems currently never
+     * set for methods at this time, so not implementing just modifying
+     * args[-1] here.
+     */
+    npy_intp len_kwargs = kwnames != NULL ? PyTuple_GET_SIZE(kwnames) : 0;
+    size_t original_arg_size = (len_args + len_kwargs) * sizeof(PyObject *);
+
+    if (NPY_UNLIKELY(len_args + len_kwargs > NPY_MAXARGS)) {
+        new_args = (PyObject **)PyMem_MALLOC(original_arg_size + sizeof(PyObject *));
+        if (new_args == NULL) {
+            /*
+             * If this fails Python uses `PY_VECTORCALL_ARGUMENTS_OFFSET` and
+             * we should probably add a fast-path for that (hopefully almost)
+             * always taken.
+             */
+            return PyErr_NoMemory();
+        }
     }
-    Py_INCREF(self);
-    PyTuple_SET_ITEM(sargs, 0, (PyObject *)self);
-    for (i = 0; i < n; ++i) {
-        PyObject *item = PyTuple_GET_ITEM(args, i);
-        Py_INCREF(item);
-        PyTuple_SET_ITEM(sargs, i+1, item);
+    else {
+        /* Almost guaranteed a stack allocation is enough. */
+        new_args = alloca(original_arg_size + sizeof(PyObject *));
     }
 
-    /* Call the function and return */
-    ret = PyObject_Call(forwarding_callable, sargs, kwds);
-    Py_DECREF(sargs);
-    return ret;
+    new_args[0] = self;
+    memcpy(&new_args[1], args, original_arg_size);
+    PyObject *res = PyObject_Vectorcall(callable, new_args, len_args+1, kwnames);
```

Comment:
For a moment, I wondered if there was a point to now set the `PY_VECTORCALL_ARGUMENTS_OFFSET` flag to indicate `self` can be overwritten, but I guess there is little point since this just goes out to python anyway.

Anyway, happy to just go with this!

Reply:
Here you would also indicate that `new_args[-1]` can be, so I don't think its valid.

Reply:
I think if one were to use it, one would have to pass in `new_args+1`. But to be honest, I'm still not sure that I really understand how it is supposed to work.

Reply:
:shrug:, for `PyObject_VectorcallMethod()` the flag is lost along way internal to Python when calling an unbound method (it could be restored in the end in theory, but is not and the information was dropped).

I would think it does its job somewhere (`tp_vectorcall` does get it) and I am sure _bound_ methods use it to insert the `self` (but then don't forward it here either, even though they could).

It would make sense for us getting the flag almost always here, but just flipping the switch may break the stable API that got away with forgetting about it, so not sure its worth pursuing...

Reply:
Anyway, I think it makes quite a lot of sense. The one likely thing here we could potentially ask from Python:  It looks like `args[-1]` is *always* `self` (which isn't surprising).

Frankly, as long as PyPy does that too, we could just abuse it, but it would maybe be better to ask for it to be documented to be true.